netlink/conntracker.go

```diff
@@ -0,0 +1,266 @@
+// +build linux
+
+package netlink
+
+import (
+	"context"
+	golog "log"
+	"net"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+
+	ct "github.com/florianl/go-conntrack"
+)
+
+// Conntracker is a wrapper around go-conntracker that keeps a record of all connections in user space
+type Conntracker interface {
+	GetConntrackEntryForConn(ip string, port uint16) *IPTranslation
+	Close()
+}
+
+type connKey struct {
+	ip   string
+	port uint16
+}
+
+type realConntracker struct {
+	nfct  *ct.Nfct
+	l     *sync.Mutex
+	state map[connKey]*IPTranslation
+
+	statsTicker   *time.Ticker
+	compactTicker *time.Ticker
+	stats         struct {
+		gets               int64
+		getTimeTotal       int64
+		registers          int64
+		registersTotalTime int64
+	}
+}
+
+func NewConntracker() (Conntracker, error) {
+	nfct, err := ct.Open(&ct.Config{ReadTimeout: 10 * time.Millisecond, Logger: golog.New(os.Stdout, "go-conntrack", 0)})
+	if err != nil {
+		return nil, err
+	}
+
+	ctr := &realConntracker{
+		nfct:          nfct,
+		l:             &sync.Mutex{},
+		statsTicker:   time.NewTicker(time.Second * 10),
+		compactTicker: time.NewTicker(time.Hour),
+		state:         make(map[connKey]*IPTranslation),
+	}
+
+	// seed the state
+	sessions, err := nfct.Dump(ct.Ct, ct.CtIPv4)
+	if err != nil {
+		return nil, err
+	}
+	ctr.registerUnlocked(sessions)
+
+	sessions, err = nfct.Dump(ct.Ct, ct.CtIPv6)
+	if err != nil {
+		// this is not fatal because we've already seeded with IPv4
+		log.Errorf("Failed to dump IPv6")
+	}
+	ctr.registerUnlocked(sessions)
+
+	go ctr.emitStats()
+	go ctr.compact()
+
+	nfct.Register(context.Background(), ct.Ct, ct.NetlinkCtNew|ct.NetlinkCtExpectedNew|ct.NetlinkCtUpdate, ctr.register)
+	nfct.Register(context.Background(), ct.Ct, ct.NetlinkCtDestroy, ctr.unregister)
+
+	log.Infof("initialized conntrack")
+
+	return ctr, nil
+}
+
+func (ctr *realConntracker) GetConntrackEntryForConn(ip string, port uint16) *IPTranslation {
+	then := time.Now().UnixNano()
+
+	ctr.l.Lock()
+	defer ctr.l.Unlock()
+
+	result := ctr.state[connKey{ip, port}]
+
+	now := time.Now().UnixNano()
+	atomic.AddInt64(&ctr.stats.gets, 1)
+	atomic.AddInt64(&ctr.stats.getTimeTotal, now-then)
+	return result
+}
+
+func (ctr *realConntracker) Close() {
+	ctr.statsTicker.Stop()
+	ctr.compactTicker.Stop()
+	ctr.nfct.Close()
+}
+
+func (ctr *realConntracker) registerUnlocked(sessions []ct.Conn) {
```

Comment:
Why is this called `registerUnlocked`?

Reply:
perhaps it can be renamed to `loadInitialState`?

Reply:
Looks like you realized. It's called registerUnlocked because the lock isn't held when it's called, so it's up to the caller (NewConntracker  to handle locking)

I like `loadInitialState` better though